libraries/tock-register-interface/src/register_traits/mod.rs

```diff
@@ -0,0 +1,65 @@
+// Licensed under the Apache License, Version 2.0 or the MIT License.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// Copyright Tock Contributors 2024.
+// Copyright Google LLC 2024.
+
+use crate::{ArrayDataType, DataType, ScalarDataType};
+
+mod read;
+mod write;
+
+pub use read::Read;
+pub use write::Write;
+
+/// Trait implemented by all register types. Note that the Register
+/// implementations only point to the real register (e.g. they are pointer or
+/// reference types), hence why they may be copied.
+///
+/// Registers are further divided into `ArrayRegister`s and `ScalarRegister`s.
+pub trait Register: Copy {
+    type DataType: DataType;
+}
+
+/// A register that can be read, but which is not memory-safe to read.
+pub trait UnsafeRead: Register {
+    /// # Safety
+    /// Reading this register has hardware-specific safety requirements which
+    /// the caller must comply with.
+    unsafe fn read(self) -> <Self::DataType as DataType>::Value
+    where
+        Self::DataType: ScalarDataType;
+
+    /// Read from an unsafe array register. Instead of using `read_at_unchecked`
+    /// directly, callers are encouraged to call `get()` to get an
+    /// `ArrayElement` pointing at the register, then invoke `read` on that.
+    ///
+    /// # Safety
+    /// `index` must be less than `Self::LEN`.
+    /// Reading this register has hardware-specific safety requirements which
+    /// the caller must comply with.
+    unsafe fn read_at_unchecked(self, index: usize) -> <Self::DataType as DataType>::Value
+    where
+        Self::DataType: ArrayDataType;
+}
+
+/// A regsiter that can be written, but which is not memory-safe to write.
+pub trait UnsafeWrite: Register {
+    /// # Safety
+    /// Writing this register has hardware-specific safety requirements which
+    /// the caller must comply with.
+    unsafe fn write(self, value: <Self::DataType as DataType>::Value)
+    where
+        Self::DataType: ScalarDataType;
+
+    /// Write to an unsafe array register. Instead of using `write_at_unchecked`
+    /// directly, callers are encouraged to call `get()` to get an
+    /// `ArrayElement` pointing at the register, then invoke `read` on that.
+    ///
+    /// # Safety
+    /// `index` must be less than `Self::LEN`.
+    /// Writing this register has hardware-specific safety requirements which
+    /// the caller must comply with.
+    unsafe fn write_at_unchecked(self, index: usize, value: <Self::DataType as DataType>::Value)
+    where
+        Self::DataType: ArrayDataType;
+}
```

Comment:
I'd prefer if we, for this particular PR, remove the content (i.e. methods) contained in these traits. I'm not strongly opposed to introducing the notion of "safety" into the register interface itself, but it's a can of worms that I think we should be carefully discussing in a follow-up PR. We do have some good arguments as to why our current approach of "unsafe construction, safe usage" as discussed in #3244 is at least sufficient, if not preferable.

In short, with this PR merged we'd still be able to specify that a register is `UnsafeRead` or `UnsafeWrite`, but those registers would -- for the time being -- not expose any useful interfaces.

We should then reason about when exactly a register must be marked unsafe, which operations it should support, and how those operations are implemented. In particular, we may want an `UnsafeRead` or `UnsafeWrite` to not be traits with first-class access methods (such as the current `read` / `write`), but simply have an unsafe `access()` method that then returns a reference to something that implements `Read` or `Write`, and as such having the same API surface as the safe counterparts.

Reply:
Requiring that all _safe_ usage of all of a peripheral's _register_ is correct at the point of the peripheral's construction results in facile load-bearing `unsafe` blocks and should be avoided. Consider that future edits to safe code will rarely check that `unsafe` preconditions are being upheld, so load-bearing safe code should be kept as local as possible.

> we may want an UnsafeRead or UnsafeWrite to not be traits with first-class access methods... simply have an unsafe access() method that then returns a reference to something that implements...

This is morally equivalent to `unsafe { &*ptr }`; it's not in any way mutually exclusive with locally-`unsafe` reads/writes. I would also recommend people prefer the locally-`unsafe` read/writes as the `unsafe` block requires local reasoning about state.

A similar option with a similar caveat would be to have the trait method to retrieve the register type from the peripheral type be made `unsafe` for unsafe registers. That would block this PR on changing the safety of the register access and I'd suggest a syntax more like `{ unsafe Read, unsafe Write }`.

One advantage of that is that [`clippy::missing_safety_doc`](https://rust-lang.github.io/rust-clippy/master/#/missing_safety_doc) can be made to enforce that `# Safety` doc strings are present for register fields that contain `UnsafeRead` or `UnsafeWrite`, instead of having a generic `# Safety 🤷🏻 ` section in the trait.

However, a disadvantage is it's no longer clear how to allow for registers that are unsafe to write + safe to read or vice versa, along with the safety caveat.

Overall I'd still prefer some way for peripheral macro users to document safety conditions that will be raised at the point where the UB may be exercised.

Reply:
I don't think I'm opposed to anything you've written. It's just that I'd like to defer having exactly this discussion to a follow-up PR.

Whether we want unsafe methods that read/write directly, or ones that yield another ephemeral reference that can then be used safely seems like a mostly stylistic question. While I don't disagree with the `unsafe { &*ptr }` argument above, it would be nice if we could avoid duplication of `Read` and `UnsafeRead` trait implementations, and it can also be nice to be able to "lease out" a safe reference to something that's generally `UnsafeRead` or `UnsafeWrite` when the `unsafe` code can assert that the respective operations are safe temporarily, for the lifetime of the handed-out reference.

Still, those are discussions that arguably we ought to have in a follow-up PR. This is complex enough already without diving in to these interface discussions. They're also orthogonal -- this PR is quite nice as it changes little about the `Read`/`Write` (ex `Readable`/`Writeable`) interfaces. In line with that, let's defer introduction of the `UnsafeRead` etc. traits to a follow-up PR.